File: src/SpaceGame/Systems/shooting_system.py
from Systems.system import System
import math
import time
import logging


class ShootingSystem(System):

    mandatory = ["shooting", "position",
                 "velocity", "rotation", "weapon", "physics_update"]
    optional = ["player_controlled"]
    # handles = ["shooting"]

    def __init__(self, node_factory, weapons):
        super().__init__(node_factory)
        self.weapons = weapons

    def handle(self, node):
        node.add_or_attach_component("shooting_vars", {})

        weapon_fn = self.weapons[node.weapon.type]

        inputs = node.shooting.inputs
        firing_rate = node.weapon.firing_rate

        now = time.time() * 1000
        dt_last_update = now - node.shooting_vars.last_update

        running_time = min(node.shooting_vars.residual_cooldown + dt_last_update,
                           firing_rate) if node.shooting_vars.residual_cooldown is not None else firing_rate
        logging.info(running_time)
        total_time = 0
        bullets_fired = node.shooting_vars.bullets_fired

        for inp in inputs:
            dt = inp["dt"]
            if inp["shooting"]:
                if running_time + dt >= firing_rate:
                    logging.info("firing shots")
                    weapon_fn(self.node_factory, node,
                              total_time, bullets_fired, inp["shooting"], last_update = node.shooting_vars.last_update)
                    running_time -= firing_rate
                    bullets_fired += 1
            else:
                # if running_time + dt >= firing_rate:
                weapon_fn(self.node_factory, node,
                          total_time, bullets_fired, inp["shooting"])
            running_time += dt
            total_time += dt

        node.shooting_vars.last_update = time.time() * 1000
        node.shooting_vars.bullets_fired = bullets_fired
        node.shooting_vars.residual_cooldown = running_time

File: src/SpaceGame/static/js/ShootingSystem.js

var ShootingSystem = (function() {
	var manditory = ['shooting', 'velocity', 'rotation', 'position', 'weapon'];
	var optional = [];
	var handles = [];

	function ShootingSystem(node_factory, textures, weapons) {
		this.node_factory = node_factory;
		this.textures = textures;
		this.weapons = weapons
	}

	ShootingSystem.prototype.process = function() {
		var nodes = this.node_factory.create_node_list(manditory, optional);
		for(let i = 0; i < nodes.length; i++) {
			let node = nodes[i];
			this.handle(node);
			this.cleanup(node);
		}

	};

	ShootingSystem.prototype.cleanup = function(node) {
		for(let i = 0; i < handles.length; i++) {
			var comp = handles[i];
			node.delete_component(comp);
		}
	}

	

	ShootingSystem.prototype.handle = function(node) {
		node.add_or_attach("shooting_vars");
 		let weapon_fn = this.weapons[node.weapon.type];
		let input = node.shooting.input;
		let firing_rate = node.shooting.firing_rate;

		let now = Date.now();
		let dt_last_time = now - node.shooting_vars.last_update

		let running_time = node.shooting_vars.residual_cooldown == null ? firing_rate : Math.min(node.shooting_vars.residual_cooldown + dt_last_time, firing_rate);
		
		// console.log("Running time", running_time);

		let bullets_fired = node.shooting_vars.bullets_fired;

		let dt = input.dt;
		if(running_time + dt >= firing_rate) {
			weapon_fn(this.node_factory, node, bullets_fired, this.textures)
			running_time -= firing_rate;
			bullets_fired++;
		}

		
		running_time += dt;

		node.shooting_vars.last_update = Date.now();
		node.shooting_vars.bullets_fired = bullets_fired;
		node.shooting_vars.residual_cooldown = running_time;

		
	}

	return ShootingSystem; 
})();

var ShootingSink = (function() {
	var manditory = ['shooting'];
	var optional = [];
	var handles = ['shooting'];

	function ShootingSink(node_factory) {
		this.node_factory = node_factory;
	}

	ShootingSink.prototype.process = function() {
		var nodes = this.node_factory.create_node_list(manditory, optional);
		for(let i = 0; i < nodes.length; i++) {
			let node = nodes[i];
			this.cleanup(node);
		}

	};

	ShootingSink.prototype.cleanup = function(node) {
		for(let i = 0; i < handles.length; i++) {
			var comp = handles[i];
			node.delete_component(comp);
		}
	}

	return ShootingSink; 
})();